### ISSUE TITLE
Model training non-reproducible when run on the same machine

### ISSUE THREAD

**Issue by matt-graham** [high-dimensional/3d_very_deep_vae#13]
Currently we do not get consistent results when running the model training scripts multiple times on the same machine. As an example, two consecutive training runs of `VeryDeepVAE_32x32x32/run_autoencoder.py` (terminating after two epochs) with the Biobank data gives the following per-epoch summary output

__Run 1__

```
# Epoch 1

Approx loss ELBO: -54055.5202, MSE: 0.0139, KL: 4017.0626
Average (non-NaN) grad norm (bottom_up_graph_1): 252088.0
Average (non-NaN) grad norm (top_down_graph_latents): 510992.0
Average (non-NaN) grad norm (top_down_graph_mu): 445208.0
Validation ELBO: -49299.1699, MSE: 0.0129, KL: 11052.0241

# Epoch 2

Approx loss ELBO: -71067.5572, MSE: 0.0015, KL: 6708.3129
Average (non-NaN) grad norm (bottom_up_graph_1): 305800.0
Average (non-NaN) grad norm (top_down_graph_latents): 812551.0
Average (non-NaN) grad norm (top_down_graph_mu): 1090513.0
Validation ELBO: -39353.5740, MSE: 0.0103, KL: 14016.7032
```

__Run 2__

```
# Epoch 1

Approx loss ELBO: -54445.0666, MSE: 0.0133, KL: 4237.0158
Average (non-NaN) grad norm (bottom_up_graph_1): 289860.0
Average (non-NaN) grad norm (top_down_graph_latents): 359009.0
Average (non-NaN) grad norm (top_down_graph_mu): 374836.0
Validation ELBO: -25294.5948, MSE: 0.0116, KL: 11051.1262

# Epoch 2
Approx loss ELBO: -71302.1374, MSE: 0.0015, KL: 6558.0556
Average (non-NaN) grad norm (bottom_up_graph_1): 301641.0
Average (non-NaN) grad norm (top_down_graph_latents): 564254.0
Average (non-NaN) grad norm (top_down_graph_mu): 824257.0
Validation ELBO: -29273.9498, MSE: 0.0086, KL: 15786.1976
```

The NumPy and in-built  `random` module pseudo-random number generator seeds are currently fixed in 

https://github.com/r-gray/3d_very_deep_vae/blob/6df8282de72e3d21cae66f9ce0eba6841bff2534/verydeepvae/orchestration/training_script_vae_new.py#L22-L30

To check if the non-determinism shown above was a result of the PyTorch random number generator seed not being set and not having called `torch.use_deterministic_algorithms(True)`  as recommended in the [PyTorch randomness documentation page](https://pytorch.org/docs/stable/notes/randomness.html) I tried re-running the training with the following lines added immediately after the above

```Python
torch.manual_seed(random_seed)
torch.use_deterministic_algorithms(True)
```

with 3 single epoch runs on the same machine giving outputs

__Run 3__

```
Approx loss ELBO: -45773.7480, MSE: 0.0247, KL: 4458.3014
Average (non-NaN) grad norm (bottom_up_graph_1): 525948.0
Average (non-NaN) grad norm (top_down_graph_latents): 1247905.0
Average (non-NaN) grad norm (top_down_graph_mu): 870809.0
Validation ELBO: -57576.0451, MSE: 0.0201, KL: 7771.3412
```

__Run 4__

```
Approx loss ELBO: -46713.3200, MSE: 0.0246, KL: 3984.8684
Average (non-NaN) grad norm (bottom_up_graph_1): 548423.0
Average (non-NaN) grad norm (top_down_graph_latents): 1184667.0
Average (non-NaN) grad norm (top_down_graph_mu): 786363.0
Validation ELBO: -53179.4228, MSE: 0.0184, KL: 6106.6914
```

__Run 5__

```
Approx loss ELBO: -45951.0240, MSE: 0.0308, KL: 3022.3111
Average (non-NaN) grad norm (bottom_up_graph_1): 457073.0
Average (non-NaN) grad norm (top_down_graph_latents): 852282.0
Average (non-NaN) grad norm (top_down_graph_mu): 576341.0
Validation ELBO: -57818.2202, MSE: 0.0189, KL: 5392.6404
```

which again show a lot of variability across the runs. It's not clear if this is due to non-seeded random generators still being used somewhere or due to some of the operations remaining non-determinstic when running on a GPU.
